### PR TITLE
csharp sqli pattern

### PR DESCRIPTION
**Pull request by gdraperi**
Hello,

I have just finished a rule for detecting SQLI in CSharp.

I am not totally satisfied because it seems not detecting when the issue is located in a parameter of a constructor.

I would need help to make it work.

Best,

Gregory

### CODE REVIEW

**Comment by SPoint42** [comment]
I just test your rules against a whole number of file where I have SQL-I in C#-.NET. Some of them was triggers, some not. I think we need to update the rules to detect more basic SQLI like this file (not detected with this version of th rules) before putting in the repo

```csharp
public string GetNonSensitiveDataById()
{
	using (SqlConnection connection = new SqlConnection(_configuration.GetValue<string>("ConnectionString")))
	{
		connection.Open();
		SqlCommand command = new SqlCommand($"SELECT * FROM NonSensitiveDataTable WHERE Id = {Request.Query["id"]}", connection);
		using (var reader = command.ExecuteReader())
		{
			if (reader.Read())
			{
				string returnString = string.Empty;
				returnString += $"Name : {reader["Name"]}. ";
				returnString += $"Description : {reader["Description"]}";
				return returnString;
			}
			else
			{
				return string.Empty;
			}
		}
	}
}

```
I just got the capability to audit some CSharp project and will see how to improove your rules

**Comment by gdraperi** [comment]
Hello, thanks for the feedback. I agree. My corpus was small so I started with a small subset of patterns. Tell me how you want to proceed.

File: csharp/lang/security/sqli/csharp_sqli.yaml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    - pattern: |
+        $T $X(...,$SQL,...)
+        {
+         ...
+         $PATTERN $SQL = new $PATTERN($S,...);
+         ...
+        }
+  - pattern-not: |
+      $S.$PATTERN = "..." + "...";
+
```
Comment (SPoint42) [inline csharp/lang/security/sqli/csharp_sqli.yaml]: add this here : 

```suggestion
- pattern-not-inside: |
          ...
          $S.Parameters
```

File: csharp/lang/security/sqli/csharp_sqli.yaml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+         $S.$PATTERN = $SQL;
+         ...
+        }
+    - pattern: |
+        $T $X(...,$SQL,...)
+        {
+         ...
+         $PATTERN $SQL = new $PATTERN($S,...);
+         ...
+        }
```
Comment (SPoint42) [inline csharp/lang/security/sqli/csharp_sqli.yaml]:
```suggestion
           }
       - pattern: |
              $PATTERN $S = new $PATTERN(...);
```

**Comment by SPoint42** [comment]
I update your yaml file and seems to work on my sample and yours.

**Review by SPoint42 — CHANGES_REQUESTED** [review]

File: csharp/lang/security/sqli/csharp_sqli.yaml [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  - pattern-not: |
+      $S.$PATTERN = "..." + "...";
+
+
+  message: |
+    Detected a formatted string in a SQL statement. This could lead to SQL injection if variables in the SQL statement are not properly sanitized.
+    Use a prepared statements instead. You can obtain a PreparedStatement using 'SqlCommand' and 'SqlParameter'.
+  languages: [csharp]
+  severity: WARNING
```
Comment (SPoint42) [inline csharp/lang/security/sqli/csharp_sqli.yaml]:
```suggestion
  severity: ERROR
    metadata:
    owasp: 'A1: Injection'
    cwe: >-
      CWE-89: Improper Neutralization of Special Elements used in an SQL
      Command ('SQL Injection')
```

**Review by SPoint42 — CHANGES_REQUESTED** [review]

File: csharp/lang/security/sqli/csharp_sqli.yaml (hunk 3 above)
Comment (gdraperi) [reply csharp/lang/security/sqli/csharp_sqli.yaml]: Thanks for the feedback. I added all your suggestions. Really appreciated.

**Comment by gdraperi** [comment]
Hello ,

The pre-commits works now but it seems the others actions fail.

The error message is quite obscure and I am not sure I can do something

```
Error: ROR] `/root/.local/bin/semgrep --skip-unknown-extensions --disable-nosem --json --autofix --dryrun --time --config yaml/semgrep/duplicate-id.yaml --config yaml/semgrep/duplicate-pattern.yaml --config yaml/semgrep/unsatisfiable.yaml --config yaml/semgrep/metadata-cwe.yaml --config yaml/semgrep/metadata-owasp.yaml --config yaml/semgrep/metadata-references.yaml --config yaml/semgrep/empty-message.yaml --config yaml/semgrep/missing-message-field.yaml --config yaml/semgrep/missing-language-field.yaml --config yaml/semgrep/metadata-technology.yaml --config yaml/semgrep/metadata-category.yaml --config yaml/semgrep/multi-line-message.yaml --config yaml/semgrep/slow-pattern-top-ellipsis.yaml --enable-metrics --debug -o /tmp/tmp0b2etp0z csharp/lang/security/sqli/csharp_sqli.cs csharp/lang/security/sqli/csharp_sqli.yaml` failed with exit code 2

This is an internal error, please file an issue at https://github.com/returntocorp/semgrep-action/issues/new/choose
and include any log output from above.
```

**Comment by minusworld** [comment]
OK, thanks. Let me take a look.

**Review by minusworld — APPROVED** [review]

**Comment by minusworld** [comment]
@gdraperi @SPoint42 Thanks to you both! Merging this in. If there are issues with it going forward, feel free to open a new PR with changes.

Thanks again!! :D

**Pull request merged**